agent/consul/watch/server_local_test.go

```diff
@@ -0,0 +1,424 @@
+		Run(func(_ mock.Arguments) {
+			// Cancel the context so that the memdb WatchCtx call will error.
+			close(abandonCh)
+		})
+
+	idx, result, err := ServerLocalBlockingQuery(
+		context.Background(),
+		provider.getStore,
+		8,
+		true,
+		provider.query,
+	)
+	// The internal cancellation error should not be propagated.
+	require.NoError(t, err)
+	require.EqualValues(t, 4, idx)
+	require.Equal(t, &testResult{value: "foo"}, result)
+}
+
+func TestServerLocalNotify_Validations(t *testing.T) {
+	provider := newMockStoreProvider(t)
+
+	type testCase struct {
+		ctx      context.Context
+		getStore func() *MockStateStore
+		query    func(memdb.WatchSet, *MockStateStore) (uint64, *testResult, error)
+		notify   func(context.Context, string, *testResult, error)
+		err      error
+	}
+
+	cases := map[string]testCase{
+		"nil-context": {
+			getStore: provider.getStore,
+			query:    provider.query,
+			notify:   provider.notify,
+			err:      errNilContext,
+		},
+		"nil-getStore": {
+			ctx:    context.Background(),
+			query:  provider.query,
+			notify: provider.notify,
+			err:    errNilGetStore,
+		},
+		"nil-query": {
+			ctx:      context.Background(),
+			getStore: provider.getStore,
+			notify:   provider.notify,
+			err:      errNilQuery,
+		},
+		"nil-notify": {
+			ctx:      context.Background(),
+			getStore: provider.getStore,
+			query:    provider.query,
+			err:      errNilNotify,
+		},
+	}
+
+	for name, tcase := range cases {
+		t.Run(name, func(t *testing.T) {
+			err := ServerLocalNotify(tcase.ctx, "test", tcase.getStore, tcase.query, tcase.notify)
+			require.ErrorIs(t, err, tcase.err)
+		})
+	}
+}
+
+func TestServerLocalNotify(t *testing.T) {
+	ctx, cancel := context.WithCancel(context.Background())
+	t.Cleanup(cancel)
+
+	abandonCh := make(chan struct{})
+
+	store := NewMockStateStore(t)
+	store.On("AbandonCh").
+		Return(closeChan(abandonCh)).
+		Times(3)
+
+	provider := newMockStoreProvider(t)
+	provider.On("getStore").
+		Return(store).
+		Times(3)
+	provider.On("query", mock.Anything, store).
+		Return(uint64(4), &testResult{value: "foo"}, nil).
+		Once()
+	provider.On("notify", ctx, "test", &testResult{value: "foo"}, nil).Once()
+	provider.On("query", mock.Anything, store).
+		Return(uint64(6), &testResult{value: "bar"}, nil).
+		Once()
+	provider.On("notify", ctx, "test", &testResult{value: "bar"}, nil).Once()
+	provider.On("query", mock.Anything, store).
+		Return(uint64(7), &testResult{value: "baz"}, context.Canceled).
+		Run(func(mock.Arguments) {
+			cancel()
+		})
+
+	err := ServerLocalNotify(ctx, "test", provider.getStore, provider.query, provider.notify)
+	require.NoError(t, err)
+
+	// wait for the context cancellation which will happen when the "query" func is run the third time
+	<-ctx.Done()
+	// wait another 100ms so that when the mock expectations are checked we would have had time to pick
+	// up on any extra invocations if the notification wasn't cancelled successfully.
+	time.Sleep(100 * time.Millisecond)
```

Comment:
Totally not a comment on your approach here, but I really wish there were a way to achieve this without sleeping, that didn't introduce extra complexity (like threading in a done channel) 🤔

Reply:
I threaded through a done function (for package internal use only) and then modified the tests to pass that in and wait for the cancellation to finish.